draft-gpew-priv-ppm.md

```diff
@@ -947,26 +947,50 @@ aggregator has validated and which fall in the batch interval of the request.
 
 Using a client-provided report multiple times within a single batch, or using
 the same report in multiple batches, may allow a server to learn information
-about the client's measurement, violating the privacy property of PPM. To
-prevent such replay attacks, this specification defines a total ordering on
-reports that aggregators can use to ensure that reports are aggregated once.
-
-Aggregate requests are ordered as follows: We say that a report `R2` follows
-report `R1` if either `R2.time > R1.time` or `R2.time == R1.time` and `R2.nonce
-> R1.nonce`. If `R2.time < R1.time`, or `R2.time == R1.time` but `R2.nonce <=
-R1.nonce`, then we say that `R2` does not follow `R1`.
-
-To prevent replay attacks, each aggregator ensures that each report it
-aggregates follows the previously aggregated report. To prevent the adversary
-from tampering with the ordering of reports, honest clients incorporate the
-ordering-sensitive parameters `(time, nonce)` into the AAD for HPKE encryption.
-Note that this strategy may result in dropping reports that happen to have the
-same timestamp and nonce value.
+about the client's measurement, violating the privacy goal of PPM. To
+prevent such replay attacks, this specification requires the aggregators to
+detect and filter out replayed reports.
+
+To detect replay attacks, each aggregator keeps track of the set of nonces
+pertaining to reports that were previously processed for a given task. If the
+leader receives a report from a client whose nonce is in this set, it simply
+ignores it. Similarly, a helper ignores an encrypted input share whose nonce is
+in this set.
+
+A malicious aggregator may attempt to force a replay by replacing the nonce
+generated by the client with a nonce its peer has not yet seen. To prevent this,
+clients incorporate the nonce into the AAD for HPKE encryption, ensuring that
+the input share is only recovered if the aggregator is given the correct nonce.
+(See {{upload-request}}.)
 
 Aggregators prevent the same report from being used in multiple batches (except
 as required by the protocol) by only responding to valid collect requests, as
 described in {{batch-parameter-validation}}.
 
+The size of the nonce set is unbounded, and depending the task, can grow quite
+large. To ensure they do not need to store these indefinitely, each aggregator
+MAY at its own discretion ignore reports with timestamps that are too far in the
+past, too far in the future, or would fall in a batch interval that has already
+been aggregated (see {{batch-parameter-validation}}).
```

Comment:
It's not at your discretion whether you ignore those which have been aggregated.

Reply:
Definitely.  This is an important point on which all helpers (and the leader) need to agree.  Discretion can't come into it or aggregation will fail.

Reply:
Nice catch.